src/Model_Hydro/CPU_Hydro/CPU_Shared_FluUtility.cpp

```diff
@@ -503,24 +504,161 @@ real Hydro_CheckMinEintInEngy( const real Dens, const real MomX, const real MomY
 //
-// Parameter   :  Input : Input value
+// Parameter   :  Cons      :  Conserved variables
+//                Prim      :  Primitive variables
+//                Info      :  Additional information
+//                Input     :  Input value
+//                Passive   :  Passive scalars
+//                File      :  __FILE__
+//                Function  :  __FUNCTION__
+//                Line      :  __LINE__
+//                Show      :  true  --> Show error message
+//                             false --> Show nothing
 //
-// Return      :  true  --> Input <= 0.0  ||  >= __FLT_MAX__  ||  != itself (Nan)
-//                false --> otherwise
+// Return      :  Case1  :  true  --> Input <= 0.0  ||  >= __FLT_MAX__  ||  != itself (Nan)
+//                          false --> otherwise
+//                Case2  :  true  --> conserved variables are unphysical or the root(SRHD only) do not exist 
+//                          false --> otherwise
+//                Case3  :  true  --> primitive variables are unphysical
+//                          false --> otherwise
 //-------------------------------------------------------------------------------------------------------
 GPU_DEVICE
-bool Hydro_CheckNegative( const real Input )
+bool Hydro_CheckUnphysical( const real Cons[], const real Prim[], const real* const Input, const real Passive[],
+                            const char Info[], const char File[], const char Function[], const int Line, bool Show )
 {
 
-   if ( Input <= (real)0.0  ||  Input >= __FLT_MAX__  ||  Input != Input )    return true;
-   else                                                                       return false;
+#  ifdef SRHD
+   real Msqr, Dsqr, E_D, M_D, Temp, Discriminant;
+#  endif
+
+   if ( Cons == NULL && Prim == NULL && Input != NULL )
+   {
+      if ( *Input <= (real)0.0  ||  *Input >= __FLT_MAX__  || *Input != *Input )           goto GOTCHA;
+   }
+
+//--------------------------------------------------------------//
+//------------ only check conserved variables-------------------//
+//--------------------------------------------------------------//
+   else if ( Cons != NULL && Prim == NULL && Input == NULL )
+   {
+//    check NaN
+      if (  Cons[DENS] != Cons[DENS]
+         || Cons[MOMX] != Cons[MOMX]
+         || Cons[MOMY] != Cons[MOMY]
+         || Cons[MOMZ] != Cons[MOMZ]
+         || Cons[ENGY] != Cons[ENGY]  )                                                   goto GOTCHA;
+
+//    check +inf and -inf for MOMX/Y/Z and negative for DENS and ENGY
+      if (  (real)  TINY_NUMBER >= Cons[DENS] || Cons[DENS]  >= (real)HUGE_NUMBER
+         || (real) -HUGE_NUMBER >= Cons[MOMX] || Cons[MOMX]  >= (real)HUGE_NUMBER
+         || (real) -HUGE_NUMBER >= Cons[MOMY] || Cons[MOMY]  >= (real)HUGE_NUMBER
+         || (real) -HUGE_NUMBER >= Cons[MOMZ] || Cons[MOMZ]  >= (real)HUGE_NUMBER
+         || (real)  TINY_NUMBER >= Cons[ENGY] || Cons[ENGY]  >= (real)HUGE_NUMBER )       goto GOTCHA;
```

Comment:
As shown below, I have changed the order not only `TINY_NUMBER` but also ` -HUGE_NUMBER`. Is this okay?
```
         if (  (real) Fields[DENS] <=  TINY_NUMBER || Fields[DENS]  >= (real)HUGE_NUMBER                                               
            || (real) Fields[MOMX] <= -HUGE_NUMBER || Fields[MOMX]  >= (real)HUGE_NUMBER
            || (real) Fields[MOMY] <= -HUGE_NUMBER || Fields[MOMY]  >= (real)HUGE_NUMBER
            || (real) Fields[MOMZ] <= -HUGE_NUMBER || Fields[MOMZ]  >= (real)HUGE_NUMBER
            || (real) Fields[ENGY] <=  TINY_NUMBER || Fields[ENGY]  >= (real)HUGE_NUMBER )
```
@hyschive

Reply:
Please skip the question above. Because you have said 
> Same for other fields.

Reply:
- [x] I prefer Cons[DENS] <= TINY_NUMBER instead of TINY_NUMBER >= Cons[DENS]. Same for other fields.
- [x]  Also check passive scalars when Passive != NULL. They should be positive-definite like DENS.